### PR TITLE
[HWASAN] Emit optimization remarks

### PR DESCRIPTION
**Pull request by vitalybuka**
Similar to #88304

SelHWAsan is optimization. We may want to diagnose compiler decisions.
Remarks is the tool for that https://llvm.org/docs/Remarks.html.

### CODE REVIEW

**Comment by fmayer** [comment]
Looks fine, could you put something in the description why we want this?

**Comment by vitalybuka** [comment]
> Looks fine, could you put something in the description why we want this?

Done

**Review by fmayer — APPROVED** [review]

File: llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp [hunk 1]
```diff
@@ -1492,23 +1493,42 @@ bool HWAddressSanitizer::instrumentStack(memtag::StackInfo &SInfo,
...
+                       bool Skip) {
+  if (Skip) {
+    ORE.emit([&]() {
+      return OptimizationRemark(DEBUG_TYPE, "Skip", &F)
+             << "Skipped: F=" << ore::NV("Function", &F);
+    });
+  } else {
+    ORE.emit([&]() {
+      return OptimizationRemarkMissed(DEBUG_TYPE, "Sanitize", &F)
+             << "Sanitized: F=" << ore::NV("Function", &F);
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1506]: Why do we need this arg? We already have the F in the output either way.

Now the output looks like this

```
--- !Missed
Pass:            hwasan
Name:            Sanitize
Function:        test_retptr
Args:
  - String:          'Sanitized: F='
  - Function:        test_retptr
...

I suggest just making this `OptimizationRemarkMissed(DEBUG_TYPE, "Sanitize", &F);` Same above
```

Comment (fmayer) [reply llvm/lib/Transforms/Instrumentation/HWAddressSanitizer.cpp:1506]: Also, wouldn't it be better to have the same name for the Skip and non skip-Cases, and just distinguish by the Miss / Pass

**Pull request merged**